src/docs/rust.mdx

```diff
@@ -0,0 +1,347 @@
+---
+title: Rust Development at Sentry
+---
+
+This is a document that contains a bunch of useful resources for getting started with Rust and adhering to our coding principles.
+
+## Getting Started
+
+- A quick introduction into the Syntax for first-timers:
+  [A half-hour to learn Rust](https://fasterthanli.me/blog/2020/a-half-hour-to-learn-rust/)
+- The Rust Book, comprehensively documenting the language:
+  [The Rust Programming Language](https://doc.rust-lang.org/book/)
+- [The Async Book](https://rust-lang.github.io/async-book/)
+- Sentry employees: Join the `#discuss-rust` channel in Slack
+
+## Coding Principles
+
+### Iterators
+
+Prefer explicit iterator types over `impl Iterator` in stable, public interfaces of published crates.
+This allows to name the type in places like associated types or globals. The type name should end with `Iter` per naming convention.
+
+In addition to the standard `Iterator` trait, always consider implementing additional traits from `std::iter`:
+
+- `FusedIterator` in all cases unless there is a strong reason not to
+- `DoubleEndedIterator` if reverse iteration is possible
+- `ExactSizeIterator` if the size is known beforehand
+
+If it is exceptionally hard to write a custom iterator, then it can also be a private [newtype](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html) around a boxed iterator:
+
+```rust
+pub struct FooIter(Box<dyn Iterator<Item = Foo>>);
+
+impl Iterator for FooIter {
+    type Item = Foo;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+```
+
+<!-- ### Async-await
+
+In general we like to migrate to async-await using code, though much code exists which does not yet do this.
+Use
+During migration you may need normal functions which return futures, for their signature you should generally prefer `-> impl Future<Output = ...>` over other return types. This allows the compiler to figure out the details for callsites and avoids having to force things to be pinned or not. Pinning should only be done explicitly when part of fields in structs or similar. -->
```

Comment:
Commented this out in accordance with @jan-auer because it conflicts with the guideline on iterators above. We can rewrite it in a follow-up PR.